cmd/hercules/main.go

```diff
@@ -0,0 +1,87 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/hercules.v1"
+)
+
+func main() {
+	if len(os.Args) == 1 || len(os.Args) > 3 {
+		fmt.Fprint(os.Stderr,
+			         "Usage: hercules <path to repo or URL> [<disk cache path>]\n")
+		os.Exit(1)
+	}
+  uri := os.Args[1]
+	var repository *git.Repository
+	var err error
+	if strings.Contains(uri, "://") {
+		if len(os.Args) > 2 {
+			repository, err = git.NewFilesystemRepository(os.Args[2])
+			if err != nil {
+				panic(err)
+			}
+		} else {
+			repository = git.NewMemoryRepository()
+		}
+		err = repository.Clone(&git.CloneOptions{
+		  URL: uri,
+	  })
+	} else {
+		if uri[len(uri) - 1] == os.PathSeparator {
+			uri = uri[:len(uri) - 1]
+		}
+		if !strings.HasSuffix(uri, ".git") {
+			uri += string(os.PathSeparator) + ".git"
+		}
+		repository, err = git.NewFilesystemRepository(uri)
+	}
+	if err != nil {
+		panic(err)
+	}
+	// core logic
+	analyser := hercules.Analyser{}
+	analyser.Initialize(repository)
+	analyser.SetOnProgress(func(commit int) {
+		fmt.Fprintf(os.Stderr, "%d\r", commit)
+	})
```

Comment:
Get rid of the `analyzer.Initialize()` and the `analyzer.SetOnProgress()` methods, make the fields of `hercules.Analyzer` public and just do something like this:

```
analyser := hercules.Analyzer{
    Repository: repository,
    OnProgress: func(commit int) {
        fmt.Fprintf(os.Stderr, "%d\r", commit)
    },
}
```